resource-impl/src/main/java/org/nmdp/service/epitope/resource/impl/GroupResource.java

```diff
@@ -265,33 +303,48 @@ private static void addToList(List<String> l, Iterable<String> i) {
 	}
 	
 	private List<GroupView> calculateGroupProbabilities(List<GroupView> groupList) {
-	    Map<Integer, DoubleContainer> groupFreqMap = new HashMap<>();
+	    Map<Object, Map<String, Double>> groupFreqMap = new HashMap<>();
 	    for (GroupView groupView : groupList) {
-	        DoubleContainer dc = new DoubleContainer();
-	        groupFreqMap.put(groupView.getGroup(), dc);
-	        if (groupView.getRace() == null) {
-	            log.debug("race missing for group (skipping probabilities: " + groupView.getGroup());
-	            return groupList;
-	        }
+	    	if (groupView.getRace() == null) {
+	    		log.debug("race missing for group (skipping probabilities: " + groupView.getGroup());
+	    		return groupList;
+	    	}
+	    	//if (null == groupView.getGroup()) continue;
+	    	HashMap<String, Double> map = new HashMap<>();
+	        groupFreqMap.put(groupView.getGroup() == null ? groupView.getError() : groupView.getGroup(), map);
 	        for (String a : groupView.getAlleleList()) {
-	            dc.add(freqService.getFrequency(groupView.getRace(), a));
+	        	// todo fix forcing ars for frequencies if/when moving to genomic freqs
+	        	String ars = arsAlleleFilter.apply(a); 
+	        	double f = freqService.getFrequency(groupView.getRace(), ars);
+	        	map.put(ars, f);
 	        }
 	    }
-	    double total = 0; 
-	    for (DoubleContainer dc : groupFreqMap.values()) {
-	        total += dc.get();
-	    }
+	    // normalize results
+	    double total = groupFreqMap.values().stream()
+	    		.map(m -> m.values().stream()
+	    				.collect(summingDouble(d -> d)))
+	    		.collect(summingDouble(d -> d));
+
         List<GroupView> newList = new ArrayList<>();
-	    for (GroupView groupView : groupList) {
-	        double probability = groupFreqMap.get(groupView.getGroup()).get() / total;
+        for (GroupView groupView : groupList) {
+        	Map<String, Double> map = groupView.getGroup() == null 
+        			? groupFreqMap.get(groupView.getError())
+        					: groupFreqMap.get(groupView.getGroup());
+	        double probability = map.values().stream()
+	        		.collect(summingDouble(d -> d)) / total;
 	        probability = round(probability);
-	        newList.add(new GroupView(groupView.getGroup(), groupView.getRace(), probability, groupView.getAlleleList(), groupView.getError()));
+	        newList.add(new GroupView(
+	        		groupView.getGroup(), 
+	        		groupView.getRace(), 
+	        		probability, 
+	        		groupView.getAlleleList(), 
+	        		groupView.getError()));
 	    }
         return newList;
     }
 
 	private double round(double d) {
-	    return (double)Math.round(d * 1000) / 1000;
+	    return (double)Math.round(d * precision) / precision;
```

Comment:
Yep, I'm converting from double to long then back to double to round a value to specified precision.  I parameterized the precision that I express frequencies in in group and match services.  Previously, I was rounding to 10E-3 and needed to expand this.  Some alleles have frequencies that are too small and a truncated to 0, meaning that some possible match grades were being discarded as possibilities.  This didn't matter when we applied a threshold, but the new rules from Jason and Katie don't consider threshold any more.

Thanks for the review and merge -- Erik


________________________________
From: Caleb Kennedy [notifications@github.com]
Sent: Tuesday, October 06, 2015 9:11 PM
To: nmdp-bioinformatics/service-epitope
Cc: Erik Pearson
Subject: Re: [service-epitope] fix frequency handling on allele and group services, update swagger documentation (#51)


In resource-impl/src/main/java/org/nmdp/service/epitope/resource/impl/GroupResource.java<https://github.com/nmdp-bioinformatics/service-epitope/pull/51#discussion_r41345322>:

>           }
>          return newList;
>      }
>
>       private double round(double d) {
> -         return (double)Math.round(d * 1000) / 1000;
> +         return (double)Math.round(d * precision) / precision;


Never mind, floating points. Got it.

—
Reply to this email directly or view it on GitHub<https://github.com/nmdp-bioinformatics/service-epitope/pull/51/files#r41345322>.